### DISCUSSION TITLE
static files

### DISCUSSION THREAD

**Discussion by slavavs82** [kartikk221/hyper-express#72] [General]
app.use(express.static('public'))
how to implement such code in hyper-express?

**Comment by slavavs82** [comment]
import serveStatic from 'serve-static';
...
webserver.use(serveStatic(path.resolve(__dirname, 'static')));

this didn't help solve the problem. Maybe there is another solution?

**Comment by ImLunaHey** [comment]
https://github.com/kartikk221/live-directory can be used for this.

**Comment by BobbyWibowo** [comment]
Anybody knows a relatively easy-to-use actual library that doesn't involve pre-caching the actual files? A bit like what serve-static would actually do?

LiveDirectory is good for generic assets embedded into the websites
And pre-caching is actually an approach that I think makes sense for such assets

But not so much for user-generated files that can be quite massive in total

Serve-static couldn't be used as-is, because there were incompatibilities with how said library would set Content-Length header (it'd try to set raw non-string value and hyper-express doesn't like that)
Forking the library to patch those out wasn't sufficient either, since something with the way said library would pipe the stream did not really work out with hyper-express
Unfortunately, it just kinda hanged on me without anything throwing exception, so I just gave up at that point since I wasn't particularly familiar enough with the internals of either libraries

Also I'm aware that if for the purpose of serving user-generated files with quite massive total size, it'll generally be better to just use an actual reverse proxy servers like nginx and whatnot, but humor me please

**Reply by kartikk221** [reply]
@BobbyWibowo HyperExpress will be receiving its own internal set of middlewares similar to Express.js for body parsing and serving static files. These will implement essentially 100% of the API of the Express.js counter-parts but with their own additional parameters to maintain high performance. For your case specifically, the HyperExpress version of serve-static will implement lazy caching for content and be flexible for controlling which files are loaded into memory and which are streamed from disk per request.

**Reply by BobbyWibowo** [reply]
That sounds absolutely wonderful
I'll definitely be looking forward to that

Thank you so much for the hard work thus far and going forward

**Reply by mindplay-dk** [reply]
10 months on and still no way to serve a file. 😐

I found `uwebsockets-express`, which does have enough compatibility with Express to run `serve-static`.

https://www.npmjs.com/package/uwebsockets-express

So this demonstrates compatibility can be achieved.

Unfortunately, it has far too many other things missing. 😕

Documentation for `hyper-express` still [claims](https://github.com/kartikk221/hyper-express/blob/master/docs/Middlewares.md) to offer "full compatibility with most existing ExpressJS middlewares", which is wrong.

These past weeks, I've ported a small-ish Express application to 4 different frameworks, at least 3 of which claim to replace Express - not *one* of them have a production-ready static file server with proper streaming, cache headers and `range` support. *Not one.*

Sadly, there is a reason why Express is still the most popular package - regardless of being an absolutely terrible piece of software, if you need to serve files (yeah duh) to my knowledge, there still isn't any viable alternative, just lots of rabbit holes and dead ends. 😥

**Reply by kartikk221** [reply]
Hey @mindplay-dk apologize for the typo. It seems the Middlewares documentation was not fixed when the main README.md was fixed. On the main README for `hyper-express` it mentions "Limited Express.js API Compatibility Through Shared Methods/Properties".

I have plans to completely rewrite `hyper-express` in TypeScript but also add 100% Node.js native / Express compatibility in the upcoming 7.0 major change. To clarify more on why full compatibility is hard can be boiled down to two main reasons:
- uWebsockets.js is fundamentally is different in operation from Node.js http module which makes adding compatibility with certain things hard and almost has to be articificially mocked.
- Express.js makes some very odd decisions which seem almost illogical at times and make performance harder to achieve / optimize.

In regards to file serving, while this isn't the most plug and play solution, In my own work with `hyper-express`, I utilize `live-directory` to first load my static files into memory and then write up my own middleware which simply maps the incoming requests's `path` property to one of the static loaded files to serve them.

**Reply by mindplay-dk** [reply]
I hear ya, Express is a mess.

Looks like I already gave `live-directory` a shot in the [past](https://github.com/kartikk221/hyper-express/issues/124#issuecomment-1272985172). I see you've made in-memory caching optional, so that's good - and you're using weak etags for uncached files, also good. Looks like ranged requests and middleware to handle cache conditions is still missing?

I could build it myself, of course - but I'm starting a new job next week, so I have other priorities right now...

This looks like a good starting point! But I will probably just stick with Express for now. Kind of wish I had time to help. 🙂
